extension/api.js

```diff
@@ -0,0 +1,80 @@
+/* eslint no-unused-vars: ["error", { "varsIgnorePattern": "skeleton" }]*/
+/* global ExtensionAPI */
+
+const PREF_HOTFIX_DONE = "extensions.hotfix-intermediate-2018.done";
+const PREF_MASTER_PASSWORD_REQUEST = "extensions.hotfix-intermediate-2018.requestMasterPassword";
+
+this.hotfixIntermediateCert = class extends ExtensionAPI {
+  onStartup() {
+    if (Services.prefs.getBoolPref(PREF_HOTFIX_DONE, false)) {
+      // Return earlier if the hotfix has been already applied.
+      return;
+    }
+    this.applyHotFix();
+  }
+
+  async applyHotFix() {
+    let XPIDatabase;
+
+    try {
+      XPIDatabase = ChromeUtils.importESModule("resource://gre/modules/addons/XPIDatabase.sys.mjs").XPIDatabase;
+    } catch (err) {
+      console.warn("Unable to load XPIDatabase.sys.mjs", err);
+      return;
+    }
+
+    if (!XPIDatabase.verifySignatures) {
+      // If this is Firefox <63, verifySignatures was an XPIProvider method, but we are not targeting
+      // Firefox versions as old as that anyway.
+      console.warn("XPIDatabase.verifySignatures not found (Firefox version is likely too old)");
+      return;
+    }
+
+    // First, check if there's a master password on the token DB. If so,
+    // the user will need to login before we can add a certificate.
+    try {
+      let tokendb = Cc["@mozilla.org/security/pk11tokendb;1"].getService(Ci.nsIPK11TokenDB);
+      let token = tokendb.getInternalKeyToken();
+
+      if (!token.checkPassword("")) {
+        // Calling `token.login(true)` here would trigger a dialog that would be requesting
+        // the user to fill in their Master Password, that would be surprising and so we
+        // bail only do that if the user has set a pref to explicitly allow that (e.g. in
+        // case we may want to allow users to manually install the extension and manually
+        // insert their Master Password to get past this corner case).
+        if (!Services.prefs.getBoolPref(PREF_MASTER_PASSWORD_REQUEST, false)) {
+          console.warn(
+            `CertDB is locked by a MasterPassword, set ${PREF_MASTER_PASSWORD_REQUEST} to true` +
```

Comment:
Can we get rid of this config option? Affected users who for some reason don't want to update can also disable and re-enable or reinstall the addon to trigger the logic again.

I was thinking of keeping it really simple this time: add cert (return if fails), and if successful verify signatures + set pref.

WDYT?

Reply:
I'm guessing that with "can also disable and re-enable" you mean them to disable the master password, manually re-install the hotfix, and then re-enable their master password, is that right?

@Rob--W I'm personally fine with that (and to be honest on my Linux system the certificate seems to have been added just fine without unlocking the master password, I'm guessing that may not be the case mainly on Windows and/or macOS).

@willdurand you ok if I remove the pref and the log? (or remove the pref and keep a variant of the log)

Reply:
I got rid of the pref, changed the log to just mention to whoever may be looking to that log that there was master password and they could tempoarily disable that and re-install or reload the addon to ensure the fixed certificate will be added successfully.